docs/source/ops/monitoring.rst

```diff
@@ -0,0 +1,130 @@
+Monitoring
+==========
+
+This section gives a high-level overview of the PrivateStorageIO monitoring efforts.
+
+
+Goals
+`````
+
+Alerting
+  Something might break soon, so somebody needs to do something.
+
+Comparing over time or experiment groups
+  Is our service slower than it was last week? Does database B answer queries faster than database A?
+
+Analyzing long-term trends
+  How big is my database and how fast is it growing? How quickly is my daily-active user count growing?
+
+
+Introduction to our dashboards
+``````````````````````````````
+
+We have two groups of dashboards: Requests (external view, RED method) and Resources (internal view, USE method).
+
+Services and their dependencies can be visualized as a tree from external-facing to internal systems.
+
+.. figure:: DAG-of-services.svg
+    :width: 200px
+    :align: center
+    :alt: DAG of services, created using https://www.nomnoml.com/
+    :figclass: align-center
+
+    DAG of services
+
+We order our dashboards like a breadth-first-search of that tree.
+This makes it easier to understand dependencies and faster to trouble shoot when a high-latency problem on a low-level service bubbles up.
+
+
+Meaning of our metrics
+``````````````````````
+
+Google's *Monitoring Distributed Systems* book about what they call the `Four Golden Signals <https://sre.google/sre-book/monitoring-distributed-systems/#xref_monitoring_golden-signals>`_ has a great explanation and definition of useful metrics:
+
+Latency
+  Requests, but also errors take time, so don't discard them.
+
+Traffic
+  What constitutes "Traffic" depends on the nature of your system.
+
+Errors
+  (The *rate* of ) failed requests.
+
+Saturation
+  How "full" your service is.  Take action (i.e. page a human) before service degrades.
+
+"If you measure all four golden signals and page a human when one signal is problematic (or, in the case of saturation, nearly problematic), your service will be at least decently covered by monitoring."
+
+
+RED method for services ("request-scoped", "external view")
+```````````````````````````````````````````````````````````
+
+Request rate, Errors, Duration (+ Saturation?)
+
+* Instrument everything that takes time and could fail
+   * "In contrast to logging, services should instrument every meaningful number available for capture." (Peter Bourgon)
+
+* Plot 99th percentile, 50th percentile and average
+   * 50th and average should be close - else something is wrong
+   * Averages sum neatly - Service latency average should be sum of child service latencies
+
+
+USE method for resources ("resource-scoped", "internal view")
+`````````````````````````````````````````````````````````````
+
+Utilization, Saturation, Errors:
+
+* CPUs
+   * Saturation (Per core? So we catch maxxed out single cores? Would that work?)

```

Comment:
In one of the previous systems we did have per-core CPU monitoring (% used I think, which I guess is what saturation means for a CPU).  This was somewhat problematic because each of the 5 storage servers has 24 cores and that's far too many lines to have one a graph and have it remain coherent.

We're stuck with these servers for a little while so I think we'll need to do something about it for monitoring.  One idea I have is to plot max saturation value - either for each CPU or maybe just for each host.  This fails to reveal changes in multi-core saturation over time but our load expected to be pretty heavily single-core (as the Tahoe storage server is single process and single threaded).  This is just one idea that's occurred to me, I don't know if it's good or not.

Reply:
Sounds good to me!